### PR TITLE
feat(takum): add the takum_log elementary function library

### PR DESCRIPTION
**Pull request by Ravenwater**
Follow-up to #1304, and the first of the deferred items from #1297.

`takum_log` had **no math functions at all** — every template in `mathlib.hpp` binds to `takum<>`, so the logarithmic type could not be used with generic numerical code. `sqrt(takum_log<32,3>(2.0))` did not compile.

## What is actually free in a logarithmic format

`|value| = sqrt(e)^l`, so any function that scales `l` by a rational factor is exact before rounding:

| function | acts on l | exactness |
|---|---|---|
| `sqrt` | `l/2` | exact, then one rounding |
| `sqr` | `2l` | exact, then one rounding |
| `rsqrt` | `-l/2` | exact, then one rounding |
| `pow(x, n)` | `n*l` | exact for integer `n`, then one rounding |
| `reciprocal` | `-l` | exact, no rounding at all (Prop. 7) |

The alternative — decode / libm / encode, which is what `takum<>` does — rounds three times and cannot carry more than a double's 53 significand bits.

That ceiling is not hypothetical. `takum_log<64,3>` reaches `p = 59`, so the logarithmic value is carried as an exact integer pair `(N, q)` meaning `N/2^q` rather than as a double: halving `l` produces a fraction one bit **wider** than the source layout holds, which a double would discard before any rounding decision was made.

## Evidence

Measured against an oracle built from exact integer arithmetic — deliberately not `long double`, which is itself too narrow above `p = 53`:

| configuration | log-domain `sqrt` correctly rounded | double round trip |
|---|---|---|
| `takum_log<16,3>` | **100%** | 100% |
| `takum_log<24,3>` | **100%** | 100% |
| `takum_log<32,3>` | **100%** | 100% |
| `takum_log<48,3>` | **100%** | 100% |
| `takum_log<64,3>` | **100%** | **0.95%** |

Ties go to even 100% of the time. So below 48 bits this buys speed; above it, correctness.

## Two claims deliberately not made

**`exp` and `log` are not free.** Extracting `ln|x| = l/2` costs nothing, but expressing the answer back as a `takum_log` needs `l' = 2*ln(l/2)` — another logarithm. Each saves one transcendental, not two. I had previously written the opposite in #1304's description; this corrects it.

**Addition is still the weak operation.** Hyperbolic functions, `log1p`, `expm1`, `fmod` and `fma` are sums of exponentials or linear-domain operations, so they have no shortcut and are evaluated as reals.

## Codec

`takum_codec` gains `encode_fraction(c, N, q)`, a third encode entry point taking the exact pair. Purely additive — an exhaustive dump of 18 configurations (530,786 observations) is **bit-for-bit identical to main** for both variants.

## Coverage

`static/tapered/takum_log/math/mathlib.cpp` pins the exactness properties against independent references rather than against itself: `sqr(sqrt(x))` round-trips, `rsqrt` equals the exact reciprocal of `sqrt`, `pow(x,2)` equals `sqr`, `pow(x,-1)` equals `reciprocal()`.

One thing worth knowing about the libm-agreement check: it compares against the function applied to the value the type **stores**, not the literal it was built from. `takum_log` represents no integer exactly — an integer `k` is `sqrt(e)^l` only for irrational `l` — so `takum_log<32,3>(3)` holds `2.99999999`, and `floor` of it is honestly `2`. My first version of the test scored those correct results as failures.

## Validation

All 12 suites pass under gcc and clang (24/24), zero new warnings, clean under ASan+UBSan and `-DNDEBUG`, `check-ascii` green.

## Still deferred from #1297

Constant tables, cross-conversion between the variants, the `arithmetic`/`logic`/`conversion` regression directories, `ucalc` registration, native arithmetic and quire support.

## Summary by CodeRabbit

- **New Features**
  - Expanded `takum_log` support across logarithmic, exponential, power, root, rounding, fractional, trigonometric, hyperbolic, error-function, and fused multiply-add operations.
  - Added classification, minimum/maximum, navigation, and remainder-related operations.
  - Improved exact logarithmic-value handling, including conversion, scaling, rounding, overflow, underflow, and special values.
- **Tests**
  - Added comprehensive verification for accuracy, edge cases, special values, and multiple configurations.

### CODE REVIEW

File: include/sw/universal/number/takum/math/exponent.hpp [hunk 1]
```diff
@@ -31,4 +32,59 @@ takum<nbits, rbits, bt> expm1(const takum<nbits, rbits, bt>& x) {
...
+	TL result;
+	if (x.isnar()) { result.setnar(); return result; }
+	if (x.iszero()) return TL(1.0);                 // e^0 == 1
+	const double l2 = 2.0 * double(x);              // the result's logarithmic value
+	int64_t c = static_cast<int64_t>(l2);
+	if (l2 < 0.0 && static_cast<double>(c) != l2) --c;
+	double m = l2 - static_cast<double>(c);
+	if (m < 0.0) m = 0.0;
+	if (m >= 1.0) { m = 0.0; ++c; }
+	auto enc = Codec::encode_rounded(c, m);
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/takum/math/exponent.hpp]: Confirmed as undefined behaviour and fixed in e51741b6 — and the consequence was worse than an out-of-range cast.

UBSan:
```
exponent.hpp:53: runtime error: signed integer overflow:
                 -9223372036854775808 - 1 cannot be represented in type long int
takum_log<16,3> maxpos = 2.28547e+55  ->  l2 = 4.57093e+55   (INT64_MAX = 9.2e18)
```

The wrap did not merely produce a wrong magnitude, it **inverted the saturation**:

| | before | after |
|---|---|---|
| `exp(maxpos)` | `0` | `2.28547e+55` (maxpos) |
| `exp(maxneg)` | `2.28547e+55` | `0` |

Your diagnosis of why is exactly right, and it is the part I had missed: `l2` scales with the *value* of `x`, not with its logarithmic value, so it exits `int64_t` range almost immediately rather than at some extreme corner.

I took your guard and added one more case. At `rbits = 5`, `double(x)` **itself** overflows for about a third of finite encodings, so `l2` is already `+/-inf` before the comparison. The bounds check handles that correctly by construction (`inf >= cmax + 1.0`), and there is an explicit `isnan` guard ahead of it for the `inf - inf` path. Verified on `takum_log<16,5>`, where `double(maxpos)` is `inf`: `exp(maxpos)` saturates upward, `exp(maxneg)` underflows to zero, UBSan clean.

You were also right that the suite never reached it. `VerifySpecialValues` now checks both saturation directions; it reports 2 failures per configuration against the unguarded conversion and is the only suite whose verdict changes.

File: include/sw/universal/number/takum/math/fma.hpp [hunk 2]
```diff
@@ -28,4 +28,17 @@ takum<nbits, rbits, bt> fma(const takum<nbits, rbits, bt>& a, const takum<nbits,
...
+// domain but the addition is not, so this follows the linear takum and fuses in
+// double, which is wider than any takum_log the codec supports up to 64 bits.
+// ---------------------------------------------------------------------------
+
+template<unsigned nbits, unsigned rbits, typename bt>
+inline takum_log<nbits, rbits, bt> fma(const takum_log<nbits, rbits, bt>& a,
+                                       const takum_log<nbits, rbits, bt>& b,
+                                       const takum_log<nbits, rbits, bt>& c) {
+	return takum_log<nbits, rbits, bt>(std::fma(double(a), double(b), double(c)));
+}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply include/sw/universal/number/takum/math/fma.hpp:50]: Agreed and corrected in e51741b6. The claim that a double "is wider than any takum_log the codec supports up to 64 bits" was simply wrong, and I had already measured the counter-example earlier in #1304 without connecting it here: at `rbits = 5` about **34%** of finite encodings convert to infinity as a double.

Your `b == 0` case is the sharp one — `std::fma(inf, 0.0, c)` is NaN, and the constructor turns that into NaR, so a well-defined product-sum becomes a non-value.

The comment now scopes the bridge to operands and exact product-sums representable in binary64, and notes that a range-safe native path would add in the linear domain without leaving the type. Left as follow-up rather than attempted here: it needs the native arithmetic that is still outstanding from #1297.

File: static/tapered/takum_log/math/mathlib.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+	nrOfFailedTestCases += ReportTestResult(
+		VerifyIntegerPow<24, 3>(reportTestCases, 7ull), "takum_log<24,3>", "integer pow identities");
+#endif
+
+#if REGRESSION_LEVEL_4
+	nrOfFailedTestCases += ReportTestResult(
+		VerifySqrtRoundTrip<32, 3>(reportTestCases, 1031ull), "takum_log<32,3>", "sqr(sqrt(x)) round-trip");
+	nrOfFailedTestCases += ReportTestResult(
+		VerifyIntegerPow<32, 3>(reportTestCases, 1031ull), "takum_log<32,3>", "integer pow identities");
+#endif
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/tapered/takum_log/math/mathlib.cpp]: Both halves correct, and this was the finding that mattered most. Fixed in e51741b6 by adding the widths rather than trimming the comment.

You caught a gap between three things that should have agreed: the comment claimed levels 3 and 4 reach `nbits = 64`, the widest registered configuration was `takum_log<32,3>`, and the PR description reported 48- and 64-bit results. Those numbers were real — I measured them in a scratch harness using `__int128` — but I did not carry that harness into the committed suite, so the claim had no backing in the repository. That is on me.

Your second point is the more dangerous one and I want to be explicit about it: with the old `int64_t` oracle, simply registering the wide cases would have made things **worse than not testing them**. At `nbits = 64`, `d.p` reaches 59 so `Q` reaches 60, and the `Q > 60` guard would have skipped nearly every input while `ReportTestResult` printed PASS. A green line claiming correct rounding at the exact width where the double round trip fails, having compared almost nothing.

So the fix is three parts:

- The oracle uses `__int128` where available, following the precedent in `blockbinary.hpp` and `blocktype/carry.hpp`, with the width limit raised from 60 to 90.
- `takum_log<48,3>` and `<64,3>` are registered at level 4. The 64-bit case **evaluates 49,152 inputs with 0 skipped** — I checked that number rather than assuming it.
- A run that compares nothing now **fails outright** instead of reporting success, and any partial skip is reported. Where `__int128` is absent (MSVC) the wide cases are skipped with an explicit message rather than passing silently; I verified that path by forcing the fallback branch, and the rest of the suite still passes.

Both new configurations pass.

File: static/tapered/takum_log/math/mathlib.cpp [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+			auto e = Codec::decode(static_cast<uint64_t>(m));
+			wide cand_num = static_cast<wide>(e.c) * (static_cast<wide>(1) << e.p)
+			              + static_cast<wide>(e.M_bits);
+			wide diff = scaled(cand_num, e.p) - half;
+			if (diff < 0) diff = -diff;
+			if (!have || diff < best) { best = diff; have = true; }
+			if (static_cast<uint64_t>(m) == gm) mine = diff;
+		}
+		++exercised;
+		if (have && mine != best) {
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/tapered/takum_log/math/mathlib.cpp]: Correct, and this is the third time in this PR that my assertion was weaker than the claim it was meant to back. Fixed in f7efa919.

You are right that `mine == best` accepts either encoding at a midpoint, so a ties-to-ODD implementation would have passed while the PR description claimed ties-to-even.

Worth adding why this was not a corner case: halving a logarithmic value produces a fraction one bit wider than the target layout holds, so the dropped bit is a tie **roughly half the time**. The check was blind on a large fraction of its own inputs.

The oracle now records every candidate tied at the minimum distance and requires the even one — the low bit of the magnitude, since the codec breaks ties toward an even trailing field and that field occupies the low bits in every layout. A run that reaches no midpoint now fails rather than passing, on the same reasoning as the existing zero-inputs guard.

Confirmed by mutation rather than by inspection. Flipping `encode_fraction`s tie rule from even to odd:

| configuration | failures caught |
|---|---|
| `takum_log<12,3>` | 254 |
| `takum_log<16,3>` | 4,079 |
| `takum_log<20,3>` | 65,279 |
| `takum_log<32,3>` | 65,329 |
| `takum_log<48,3>` | 1,530 |
| `takum_log<64,3>` | 6,120 |

The previous form of the check caught none of them.

I did not add separate hand-built midpoint cases: the exhaustive and strided sweeps already reach thousands of midpoints per configuration, and the no-midpoint guard now makes that measurable rather than assumed. Happy to add explicit ones if you would still prefer them as documentation.

File: static/tapered/takum_log/math/mathlib.cpp [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+	// logarithm and so leaves the characteristic range almost immediately; an
+	// unguarded conversion to int64_t wrapped and inverted the answer, returning
+	// zero for exp(maxpos) and maxpos for exp(maxneg).
+	{
+		TL big(sw::universal::SpecificValue::maxpos);
+		TL small(sw::universal::SpecificValue::maxneg);
+		if (!exp(big).isnar() && !(double(exp(big)) > 1.0)) {
+			++nrOfFailedTests;
+			if (reportTestCases) std::cout << "FAIL exp(maxpos) must saturate upward, got " << double(exp(big)) << '\n';
+		}
```
[the remark this answers is not in the document]
Comment (Ravenwater) [reply static/tapered/takum_log/math/mathlib.cpp:402]: Agreed and fixed in f7efa919, taking your suggestion as written.

The `!exp(big).isnar()` guard made the assertion vacuous for precisely the value most likely to appear if the range guard is ever removed again — NaR is what an unguarded out-of-range path tends to produce. It now requires the maxpos encoding bit for bit, and computes `exp` once rather than three times.

Confirmed by mutation: making the overflow branch return NaR instead of `maxpos()` is caught, one failure per configuration.
